### PR TITLE
ENH: Finish qMRI file-level rules

### PR DESCRIPTION
**Pull request by effigies**
Partner PR to https://github.com/bids-standard/bids-specification/pull/728. If that goes in without modification, this should be able to as well.

### CODE REVIEW

File: bids-validator/bids_validator/rules/file_level_rules.json [hunk 1]
```diff
@@ -236,7 +234,7 @@
   },
 
   "fmap_TB1EPI": {
-    "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?fmap[\\/\\\\]\\1(_\\2)?(?:_acq-[a-zA-Z0-9]+)?(?:_ce-[a-zA-Z0-9]+)?(?:_rec-[a-zA-Z0-9]+)(?:_run-[0-9]+)?_echo-[0-9]+?_flip-[0-9]+?(?:_part-(imag|mag|phase|real))?_(?:@@@_field_map_type_@@@)\\.(@@@_field_map_ext_@@@)$",
+    "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?fmap[\\/\\\\]\\1(_\\2)?(?:_acq-[a-zA-Z0-9]+)?(?:_ce-[a-zA-Z0-9]+)?(?:_rec-[a-zA-Z0-9]+)(?:_run-[0-9]+)?_echo-[0-9]+?_flip-[0-9]+?(?:_inv-[0-9]+)(?:_part-(imag|mag|phase|real))?_(?:@@@_field_map_type_@@@)\\.(@@@_field_map_ext_@@@)$",
```
Comment (tsalo) [inline bids-validator/bids_validator/rules/file_level_rules.json]:
```suggestion
    "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?fmap[\\/\\\\]\\1(_\\2)?(?:_acq-[a-zA-Z0-9]+)?(?:_ce-[a-zA-Z0-9]+)?(?:_rec-[a-zA-Z0-9]+)?(?:_run-[0-9]+)?_echo-[0-9]+?_flip-[0-9]+?(?:_inv-[0-9]+)?(?:_part-(imag|mag|phase|real))?_(?:@@@_field_map_type_@@@)\\.(@@@_field_map_ext_@@@)$",
```

File: bids-validator/bids_validator/rules/file_level_rules.json [hunk 2]
```diff
@@ -245,6 +243,40 @@
     }
   },
 
+  "fmap_rf": {
+    "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?fmap[\\/\\\\]\\1(_\\2)?(?:_acq-[a-zA-Z0-9]+)?(?:_ce-[a-zA-Z0-9]+)?(?:_rec-[a-zA-Z0-9]+)(?:_run-[0-9]+)?(?:_echo-[0-9]+)?(?:_flip-[0-9]+)?(?:_inv-[0-9]+)?(?:_part-(imag|mag|phase|real))?_(?:@@@_field_map_type_@@@)\\.(@@@_field_map_ext_@@@)$",
```
Comment (tsalo) [inline bids-validator/bids_validator/rules/file_level_rules.json]:
```suggestion
    "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?fmap[\\/\\\\]\\1(_\\2)?(?:_acq-[a-zA-Z0-9]+)?(?:_ce-[a-zA-Z0-9]+)?(?:_rec-[a-zA-Z0-9]+)?(?:_run-[0-9]+)?(?:_echo-[0-9]+)?(?:_flip-[0-9]+)?(?:_inv-[0-9]+)?(?:_part-(imag|mag|phase|real))?_(?:@@@_field_map_type_@@@)\\.(@@@_field_map_ext_@@@)$",
```

File: bids-validator/bids_validator/rules/file_level_rules.json [hunk 3]
```diff
@@ -245,6 +243,40 @@
...
+        "TB1TFL",
+        "TB1RFM",
+        "RB1COR"
+      ],
+      "@@@_field_map_ext_@@@": ["nii\\.gz", "nii", "json"]
+    }
+  },
+
+  "fmap_TB1SRGE": {
+    "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?fmap[\\/\\\\]\\1(_\\2)?(?:_acq-[a-zA-Z0-9]+)?(?:_ce-[a-zA-Z0-9]+)?(?:_rec-[a-zA-Z0-9]+)(?:_run-[0-9]+)?(?:_echo-[0-9]+)?_flip-[0-9]+?_inv-[0-9]+?(?:_part-(imag|mag|phase|real))?_(?:@@@_field_map_type_@@@)\\.(@@@_field_map_ext_@@@)$",
```
Comment (tsalo) [inline bids-validator/bids_validator/rules/file_level_rules.json]:
```suggestion
    "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?fmap[\\/\\\\]\\1(_\\2)?(?:_acq-[a-zA-Z0-9]+)?(?:_ce-[a-zA-Z0-9]+)?(?:_rec-[a-zA-Z0-9]+)?(?:_run-[0-9]+)?(?:_echo-[0-9]+)?_flip-[0-9]+?_inv-[0-9]+?(?:_part-(imag|mag|phase|real))?_(?:@@@_field_map_type_@@@)\\.(@@@_field_map_ext_@@@)$",
```

File: bids-validator/bids_validator/rules/file_level_rules.json [hunk 4]
```diff
@@ -245,6 +243,40 @@
...
+    "tokens": {
+      "@@@_field_map_type_@@@": [
+        "TB1SRGE"
+      ],
+      "@@@_field_map_ext_@@@": ["nii\\.gz", "nii", "json"]
+    }
+  },
+
+  "fmap_parametric": {
+    "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?fmap[\\/\\\\]\\1(_\\2)?(?:_acq-[a-zA-Z0-9]+)?(?:_ce-[a-zA-Z0-9]+)?(?:_rec-[a-zA-Z0-9]+)(?:_run-[0-9]+)?_(?:@@@_field_map_type_@@@)\\.(@@@_field_map_ext_@@@)$",
```
Comment (tsalo) [inline bids-validator/bids_validator/rules/file_level_rules.json]:
```suggestion
    "regexp": "^[\\/\\\\](sub-[a-zA-Z0-9]+)[\\/\\\\](?:(ses-[a-zA-Z0-9]+)[\\/\\\\])?fmap[\\/\\\\]\\1(_\\2)?(?:_acq-[a-zA-Z0-9]+)?(?:_ce-[a-zA-Z0-9]+)?(?:_rec-[a-zA-Z0-9]+)?(?:_run-[0-9]+)?_(?:@@@_field_map_type_@@@)\\.(@@@_field_map_ext_@@@)$",
```

**Review by tsalo — CHANGES_REQUESTED** [review]
I don't know how important that extra `?` is, but it's also missing from the pattern for `_TB1DAM`.

**Comment by effigies** [comment]
`(?:...)` just means that the group won't be captured, e.g., you can't use a back-reference to say "find the same chunk again". `(...)?` means the group can be present or absent.

**Review by sappelhoff — COMMENTED** [review]
I tested this PR on the qMRI examples PR, and there are some errors left:

<details>
Validating dataset qmri_mp2rage/
bids-validator@1.6.2-dev.0

        1: [WARN] The Authors field of dataset_description.json should contain an array of fields - with one author per field. This was triggered because there are no authors, which will make DOI registration from dataset metadata impossible. (code: 113 - NO_AUTHORS)

        Please visit https://neurostars.org/search?q=NO_AUTHORS for existing conversations about this issue.


        Summary:             Available Tasks:        Available Modalities: 
        9 Files, 436B                                MP2RAGE               
        1 - Subject                                  defacemask            
        1 - Session                                                        


        If you have any questions, please post on https://neurostars.org/tags/bids.

Validating dataset qmri_mp2rageme/
bids-validator@1.6.2-dev.0

        1: [ERR] Files with such naming scheme are not part of BIDS specification. This error is most commonly caused by typos in file names that make them not BIDS compatible. Please consult the specification and make sure your files are named correctly. If this is not a file naming issue (for example when including files not yet covered by the BIDS specification) you should include a ".bidsignore" file in your dataset (see https://github.com/bids-standard/bids-validator#bidsignore for details). Please note that derived (processed) data should be placed in /derivatives folder and source data (such as DICOMS or behavioural logs in proprietary formats) should be placed in the /sourcedata folder. (code: 1 - NOT_INCLUDED)
                ./sub-1/fmap/sub-1_TB1map.json
                        Evidence: sub-1_TB1map.json
                ./sub-1/fmap/sub-1_TB1map.nii
                        Evidence: sub-1_TB1map.nii

        Please visit https://neurostars.org/search?q=NOT_INCLUDED for existing conversations about this issue.

        1: [WARN] The Authors field of dataset_description.json should contain an array of fields - with one author per field. This was triggered because there are no authors, which will make DOI registration from dataset metadata impossible. (code: 113 - NO_AUTHORS)

        Please visit https://neurostars.org/search?q=NO_AUTHORS for existing conversations about this issue.


        Summary:                Available Tasks:        Available Modalities: 
        19 Files, 1.55KB                                MP2RAGE               
        1 - Subject                                                           
        1 - Session                                                           


        If you have any questions, please post on https://neurostars.org/tags/bids.

Validating dataset qmri_mpm/
bids-validator@1.6.2-dev.0

        1: [ERR] Files with such naming scheme are not part of BIDS specification. This error is most commonly caused by typos in file names that make them not BIDS compatible. Please consult the specification and make sure your files are named correctly. If this is not a file naming issue (for example when including files not yet covered by the BIDS specification) you should include a ".bidsignore" file in your dataset (see https://github.com/bids-standard/bids-validator#bidsignore for details). Please note that derived (processed) data should be placed in /derivatives folder and source data (such as DICOMS or behavioural logs in proprietary formats) should be placed in the /sourcedata folder. (code: 1 - NOT_INCLUDED)
                ./sub-01/fmap/sub-01_acq-bodyMTw_RB1COR.json
                        Evidence: sub-01_acq-bodyMTw_RB1COR.json
                ./sub-01/fmap/sub-01_acq-bodyMTw_RB1COR.nii
                        Evidence: sub-01_acq-bodyMTw_RB1COR.nii
                ./sub-01/fmap/sub-01_acq-bodyPDw_RB1COR.json
                        Evidence: sub-01_acq-bodyPDw_RB1COR.json
                ./sub-01/fmap/sub-01_acq-bodyPDw_RB1COR.nii
                        Evidence: sub-01_acq-bodyPDw_RB1COR.nii
                ./sub-01/fmap/sub-01_acq-bodyT1w_RB1COR.json
                        Evidence: sub-01_acq-bodyT1w_RB1COR.json
                ./sub-01/fmap/sub-01_acq-bodyT1w_RB1COR.nii
                        Evidence: sub-01_acq-bodyT1w_RB1COR.nii
                ./sub-01/fmap/sub-01_acq-headMTw_RB1COR.json
                        Evidence: sub-01_acq-headMTw_RB1COR.json
                ./sub-01/fmap/sub-01_acq-headMTw_RB1COR.nii
                        Evidence: sub-01_acq-headMTw_RB1COR.nii
                ./sub-01/fmap/sub-01_acq-headPDw_RB1COR.json
                        Evidence: sub-01_acq-headPDw_RB1COR.json
                ./sub-01/fmap/sub-01_acq-headPDw_RB1COR.nii
                        Evidence: sub-01_acq-headPDw_RB1COR.nii
                ... and 2 more files having this issue (Use --verbose to see them all).

        Please visit https://neurostars.org/search?q=NOT_INCLUDED for existing conversations about this issue.

        1: [WARN] 'TotalReadoutTime' is greater than 10 are you sure it's expressed in seconds? (code: 5 - TOTAL_READOUT_TIME_GREATER_THAN)
                ./sub-01/fmap/sub-01_echo-1_flip-01_TB1EPI.json
                ./sub-01/fmap/sub-01_echo-1_flip-02_TB1EPI.json
                ./sub-01/fmap/sub-01_echo-1_flip-03_TB1EPI.json
                ./sub-01/fmap/sub-01_echo-1_flip-04_TB1EPI.json
                ./sub-01/fmap/sub-01_echo-1_flip-05_TB1EPI.json
                ./sub-01/fmap/sub-01_echo-1_flip-06_TB1EPI.json
                ./sub-01/fmap/sub-01_echo-1_flip-07_TB1EPI.json
                ./sub-01/fmap/sub-01_echo-1_flip-08_TB1EPI.json
                ./sub-01/fmap/sub-01_echo-1_flip-09_TB1EPI.json
                ./sub-01/fmap/sub-01_echo-1_flip-10_TB1EPI.json
                ... and 12 more files having this issue (Use --verbose to see them all).

        Please visit https://neurostars.org/search?q=TOTAL_READOUT_TIME_GREATER_THAN for existing conversations about this issue.

        2: [WARN] The Authors field of dataset_description.json should contain an array of fields - with one author per field. This was triggered based on the presence of only one author field. Please ignore if all contributors are already properly listed. (code: 102 - TOO_FEW_AUTHORS)

        Please visit https://neurostars.org/search?q=TOO_FEW_AUTHORS for existing conversations about this issue.


        Summary:                  Available Tasks:        Available Modalities: 
        108 Files, 54.84KB                                MPM                   
        1 - Subject                                       TB1EPI                
        1 - Session                                       fieldmap              


        If you have any questions, please post on https://neurostars.org/tags/bids.

Validating dataset qmri_mtsat/
bids-validator@1.6.2-dev.0

        1: [WARN] The Authors field of dataset_description.json should contain an array of fields - with one author per field. This was triggered based on the presence of only one author field. Please ignore if all contributors are already properly listed. (code: 102 - TOO_FEW_AUTHORS)

        Please visit https://neurostars.org/search?q=TOO_FEW_AUTHORS for existing conversations about this issue.


        Summary:             Available Tasks:        Available Modalities: 
        8 Files, 532B                                MTS                   
        1 - Subject                                                        
        1 - Session                                                        


        If you have any questions, please post on https://neurostars.org/tags/bids.

Validating dataset qmri_qsm/
bids-validator@1.6.2-dev.0

        1: [WARN] The Authors field of dataset_description.json should contain an array of fields - with one author per field. This was triggered based on the presence of only one author field. Please ignore if all contributors are already properly listed. (code: 102 - TOO_FEW_AUTHORS)

        Please visit https://neurostars.org/search?q=TOO_FEW_AUTHORS for existing conversations about this issue.


        Summary:             Available Tasks:        Available Modalities: 
        4 Files, 864B                                Chimap                
        1 - Subject                                                        
        1 - Session                                                        


        If you have any questions, please post on https://neurostars.org/tags/bids.

Validating dataset qmri_sa2rage/
bids-validator@1.6.2-dev.0

        1: [ERR] Files with such naming scheme are not part of BIDS specification. This error is most commonly caused by typos in file names that make them not BIDS compatible. Please consult the specification and make sure your files are named correctly. If this is not a file naming issue (for example when including files not yet covered by the BIDS specification) you should include a ".bidsignore" file in your dataset (see https://github.com/bids-standard/bids-validator#bidsignore for details). Please note that derived (processed) data should be placed in /derivatives folder and source data (such as DICOMS or behavioural logs in proprietary formats) should be placed in the /sourcedata folder. (code: 1 - NOT_INCLUDED)
                ./sub-01/fmap/sub-01_acq-td1_TB1SRGE.json
                        Evidence: sub-01_acq-td1_TB1SRGE.json
                ./sub-01/fmap/sub-01_acq-td1_TB1SRGE.nii.gz
                        Evidence: sub-01_acq-td1_TB1SRGE.nii.gz
                ./sub-01/fmap/sub-01_acq-td2_TB1SRGE.json
                        Evidence: sub-01_acq-td2_TB1SRGE.json
                ./sub-01/fmap/sub-01_acq-td2_TB1SRGE.nii.gz
                        Evidence: sub-01_acq-td2_TB1SRGE.nii.gz

        Please visit https://neurostars.org/search?q=NOT_INCLUDED for existing conversations about this issue.

        1: [WARN] The Authors field of dataset_description.json should contain an array of fields - with one author per field. This was triggered because there are no authors, which will make DOI registration from dataset metadata impossible. (code: 113 - NO_AUTHORS)

        Please visit https://neurostars.org/search?q=NO_AUTHORS for existing conversations about this issue.


        Summary:                  Available Tasks:        Available Modalities: 
        10 Files, 216.04KB                                MP2RAGE               
        1 - Subject                                                             
        1 - Session                                                             


        If you have any questions, please post on https://neurostars.org/tags/bids.

Validating dataset qmri_vfa/
bids-validator@1.6.2-dev.0

        1: [ERR] Files with such naming scheme are not part of BIDS specification. This error is most commonly caused by typos in file names that make them not BIDS compatible. Please consult the specification and make sure your files are named correctly. If this is not a file naming issue (for example when including files not yet covered by the BIDS specification) you should include a ".bidsignore" file in your dataset (see https://github.com/bids-standard/bids-validator#bidsignore for details). Please note that derived (processed) data should be placed in /derivatives folder and source data (such as DICOMS or behavioural logs in proprietary formats) should be placed in the /sourcedata folder. (code: 1 - NOT_INCLUDED)
                ./sub-01/fmap/sub-01_TB1map.json
                        Evidence: sub-01_TB1map.json
                ./sub-01/fmap/sub-01_TB1map.nii.gz
                        Evidence: sub-01_TB1map.nii.gz

        Please visit https://neurostars.org/search?q=NOT_INCLUDED for existing conversations about this issue.

        1: [WARN] The Authors field of dataset_description.json should contain an array of fields - with one author per field. This was triggered based on the presence of only one author field. Please ignore if all contributors are already properly listed. (code: 102 - TOO_FEW_AUTHORS)

        Please visit https://neurostars.org/search?q=TOO_FEW_AUTHORS for existing conversations about this issue.


        Summary:             Available Tasks:        Available Modalities: 
        8 Files, 865B                                VFA                   
        1 - Subject                                                        
        1 - Session                                                        


        If you have any questions, please post on https://neurostars.org/tags/bids.

</details>

In particular:

1.  

image


2. 

image

3. 

image


4.

image

--> some qmri datasets are passing now :tada:

**Comment by sappelhoff** [comment]
https://github.com/bids-standard/bids-validator/pull/1186/commits/f714d63185e9465c586b4b88f738fbebe85fc954 fixes the errors reported above, except for:

image

**Comment by sappelhoff** [comment]
https://github.com/bids-standard/bids-validator/pull/1186/commits/6058d9ed3b34afec61632cfefb9bc7b36bff8075 fixes the remaining issue. The auto style fix has obscured the diff, but this was the essential change:

```diff
- ?_flip-[0-9]+?_inv-[0-9]+?
+ ?(?:_flip-[0-9]+)?(?:_inv-[0-9]+)?
```

In case this fix was wrong, we can revert it ... and fix the dataset instead.

In case this fix was fine, we can merge this and then we're almost done (only warnings of two kinds remaining for the qMRI examples). @tsalo @effigies

**Comment by tsalo** [comment]
Thanks @sappelhoff! Apologies if this came up elsewhere, but are we sure that the examples are right? `TB1SRGE` requires `flip` and `inv` in the schema right now, and the [File collections](https://bids-specification.readthedocs.io/en/latest/99-appendices/10-file-collections.html#fieldmap-data) page says that `flip` and `inv` are required as well.

**Comment by sappelhoff** [comment]
> but are we sure that the examples are right? 

Not sure 😞

**Comment by sappelhoff** [comment]
>  but are we sure that the examples are right? TB1SRGE requires flip and inv in the schema right now, and the File collections page says that flip and inv are required as well.

@agahkarakuzu @Gilles86 can you shed some light on this, please?

**Comment by effigies** [comment]
I added flip/inv to the example dataset.

**Comment by agahkarakuzu** [comment]
@sappelhoff I am also working on the example datasets on my fork (https://github.com/agahkarakuzu/BIDS-examples/tree/qmri), but had some issues with syncing changes with the validator. 

As for TB1SRGE, `flip` and `inv` are required entities. Maybe I can go ahead and push all the examples first, then deal with the validator later?

**Comment by sappelhoff** [comment]
> I added flip/inv to the example dataset.

Thanks @effigies - and as @agahkarakuzu mentioned right afterwards, this was good:

> As for TB1SRGE, flip and inv are required entities. 

---

Now we just need to:
1. revert my commit that made flip and inv optional
2. and then merge this PR
3. check if the examples are passing

Perhaps we can merge @effigies's PR into the https://github.com/bids-standard/bids-examples/tree/bep001_qmri branch first, instead of `master`. Then @agahkarakuzu can put his changes on top of that ... and then finally, we merge that branch into `master`.

WDYT?

**Comment by effigies** [comment]
Stefan, please feel free to do what you think necessary. I'll make you an admin on my forks of validator and examples.

File: bids-validator/utils/type.js [hunk 5]
```diff
@@ -212,18 +215,24 @@ export default {
...
-          conditionalMatch(fmapTB1EPI, path))
+        (!path.endsWith('.json') &&
+          /* isFieldMap */
+          (conditionalMatch(fmapGre, path) ||
+            conditionalMatch(fmapPepolarAsl, path) ||
+            conditionalMatch(fmapTB1DAM, path) ||
+            conditionalMatch(fmapTB1EPI, path))) ||
+        conditionalMatch(fmaprf, path) ||
+        conditionalMatch(fmapTB1SRGE, path) ||
+        conditionalMatch(fmapparametric, path)
```
Comment (effigies) [inline bids-validator/utils/type.js:235]: Just a quick note that if this logic doesn't work, the idea is just find all files that match `isFieldMap` and aren't JSON. The original parenthesis pattern could be restored.

**Pull request merged**